File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/editor/engine/PipMaskRenderer.kt
package com.moon.videomerger.editor.engine

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import com.moon.videomerger.editor.data.PipShape
import java.io.File
import java.io.FileOutputStream
import kotlin.math.min

/**
 * 画中画形状蒙版渲染器 —— 用 Android Canvas 生成 PNG 蒙版/描边，
 * 供 ffmpeg 通过 alphamerge / overlay 实现圆角/圆形裁剪与描边。
 *
 * 为什么用 PNG 而不是 ffmpeg 的 geq：min/full 版无 drawtext 时已确立
 * Canvas 生成 PNG 的替代方案，这里沿用同一思路，逻辑可读、可单测、可与预览一致。
 */
object PipMaskRenderer {

    /** 生成形状蒙版（白色实心形状，透明背景）。用于 alphamerge 裁剪画中画。 */
    fun renderMask(context: Context, w: Int, h: Int, shape: PipShape, cornerRadius: Float): File {
        val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bmp)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            style = Paint.Style.FILL
        }
        drawShape(canvas, w.toFloat(), h.toFloat(), shape, cornerRadius, paint, 0f)
        return savePng(context, bmp, "pip_mask")
    }

    /** 生成描边框（白色描边，透明内部）。用于 overlay 叠加画中画描边。 */
    fun renderBorder(context: Context, w: Int, h: Int, shape: PipShape, cornerRadius: Float, strokeWidth: Float): File {
        val bmp = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bmp)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.WHITE
            style = Paint.Style.STROKE
            this.strokeWidth = strokeWidth
        }
        // 描边居中于形状边缘，向内缩进半个描边宽度避免被画布裁掉
        drawShape(canvas, w.toFloat(), h.toFloat(), shape, cornerRadius, paint, strokeWidth / 2f)
        return savePng(context, bmp, "pip_border")
    }

    private fun drawShape(
        canvas: Canvas,
        w: Float,
        h: Float,
        shape: PipShape,
        cornerRadius: Float,
        paint: Paint,
        inset: Float
    ) {
        when (shape) {
            PipShape.RECT -> canvas.drawRect(RectF(inset, inset, w - inset, h - inset), paint)
            PipShape.ROUNDED -> canvas.drawRoundRect(RectF(inset, inset, w - inset, h - inset), cornerRadius, cornerRadius, paint)
            PipShape.CIRCLE -> {
                val d = min(w, h) - inset * 2f
                canvas.drawCircle(w / 2f, h / 2f, d / 2f, paint)
            }
        }
    }

    private fun savePng(context: Context, bmp: Bitmap, prefix: String): File {
        val file = File(context.cacheDir, "${prefix}_${System.currentTimeMillis()}_${(Math.random() * 100000).toInt()}.png")
        FileOutputStream(file).use { out ->
            bmp.compress(Bitmap.CompressFormat.PNG, 100, out)
        }
        bmp.recycle()
        return file
    }
}

File: VideoMergeIOS/VideoMerger/Editor/Engine/PipMaskRenderer.swift
//
//  PipMaskRenderer.swift
//  VideoMerger
//
//  iOS 移植自 Android PipMaskRenderer.kt
//  画中画形状蒙版（alphamerge 用）与描边 PNG
//

import UIKit

enum PipMaskRenderer {

    /// 形状蒙版：白色形状 + 透明背景（alphamerge 用亮度替换 alpha）
    static func renderMask(width: Int, height: Int, shape: PipShape, radius: CGFloat) -> URL? {
        render(width: width, height: height, shape: shape, radius: radius, stroke: false)
    }

    /// 白色描边框
    static func renderBorder(width: Int, height: Int, shape: PipShape, radius: CGFloat, strokeWidth: Float) -> URL? {
        render(width: width, height: height, shape: shape, radius: radius, stroke: true, strokeWidth: strokeWidth)
    }

    private static func render(width: Int, height: Int, shape: PipShape, radius: CGFloat,
                               stroke: Bool, strokeWidth: Float = 0) -> URL? {
        guard width > 0, height > 0 else { return nil }
        let size = CGSize(width: width, height: height)
        let renderer = UIGraphicsImageRenderer(size: size)
        let image = renderer.image { _ in
            // 描边居中于形状边缘，向内缩进半个描边宽度避免被画布裁掉（对齐 Android）
            let inset = stroke ? CGFloat(strokeWidth) / 2 : 0
            let w = CGFloat(width), h = CGFloat(height)
            let path: UIBezierPath
            switch shape {
            case .circle:
                // ★ 内切正圆：直径取宽高中的较小者，而不是椭圆
                let d = min(w, h) - inset * 2
                path = UIBezierPath(ovalIn: CGRect(x: (w - d) / 2, y: (h - d) / 2, width: d, height: d))
            case .rounded:
                path = UIBezierPath(roundedRect: CGRect(x: inset, y: inset, width: w - inset * 2, height: h - inset * 2),
                                    cornerRadius: radius)
            default:
                path = UIBezierPath(rect: CGRect(x: inset, y: inset, width: w - inset * 2, height: h - inset * 2))
            }
            if stroke {
                UIColor.white.setStroke()
                path.lineWidth = CGFloat(strokeWidth)
                path.stroke()
            } else {
                UIColor.white.setFill()
                path.fill()
            }
        }
        guard let data = image.pngData() else { return nil }
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("pip_mask_\(Int(Date().timeIntervalSince1970 * 1000))_\(Int.random(in: 0..<9999)).png")
        do { try data.write(to: url) } catch { return nil }
        return url
    }
}
